components/scream/src/share/util/scream_utils.hpp

```diff
@@ -0,0 +1,191 @@
+// the ordering of elements (A,B,C,D,E,F,G) clearly has all G's as contiguous
+// subset, though it might not be obvious if one scrambled the entris within
+// each input group.
+// If such an ordering cannot be found, we return an empty list.
+
+// The algorithm is the following.
+//  - We proceed iteratively, trying to add one group at a time to the result.
+//  - Internally, we store the result as a list of lists (LoL), like
+//        [ [A,B], [C], [D,E,F], [G,H], [I], ... ]
+//    If the algorithm succeeds, the output of the function is a single list
+//    obtainied by splicing together all the inner lists.
+//  - We are only allowed to modify the outer list in the following ways:
+//      a) rearrange the elements in an inner list: [D,E,F]->[F,D,E]
+//      b) split an inner list into multiple lists: [[A,B,C],[D,E]]->[[A,B],[C],[D,E}]
+//      c) append a new list (that does not overlap with existing sublists):
+//        [[A,B],[C,D]] -> [[A,B],[C,D],[E,F,G]]
+//  - The 1st group is added as a single sublist, that is, given G1,..,G5 above,
+//    after the first iteration, the list of lists would be [[A,B,C]].
+//  - At the generic k-th step of the algorithm, we try to add group G to the LoL,
+//    doing the following:
+//     - compute intersection of G with each of the sublists of LoL
+//     - check that all non-empty intersections are 'contiguous'. If not, the
+//       overall algorithm failed, and we can return.
+//     - if there's a part of G that does not intersect with any sublist, then
+//       there are two scenarios:
+//        i) all intersections are at the beginning or end of LoL: good.
+//        ii) intersections are neither at the beginning nor at the end of LoL.
+//            In this case, the overall algorithm failed, and we can return.
+//     - if G has a non-empty intersection with sublist S, there are two scenarios:
+//        i) the intersection is the whole S: nothing to do.
+//        ii) the intersection is smaller than S: two sub-scenarios:
+//          *) S is the first or last sublist with non-empty intersection:
+//             split S into [intersection][rest] or [rest][intersection]
+//             (the former if it's the first intersection, the former if it's the last).
+//             This would correspond to operation (a+b) on the LoL.
+//          *) S is neither the first nor the last. In this case the overall
+//             algorithm has failed, and we can return.
+//     - if G has a portion that does not intersect with any sublist S, add it
+//       as a new sublist at the end. This corresponds to operation (c) from above.
+
+// Regardless of the order in which we process the individual lists, if there is
+// an arrangement A of T's that allows to have G1,...,Gn as a contiguous sublist
+// of A, the above algorithm is guaranteed to find it.
+
+template<typename T>
+std::list<T> contiguous_superset (const std::list<std::list<T>>& groups)
```

Comment:
I'm not sold on the name. What the fcn does is, given `lol=list<list<T>>` find an ordering of all the elements `ordered=list<T>` such that each list in `lol` is a contiguous sublist of `ordered`.

Reply:
Can you give a contextual example of how this is used?

Reply:
Take the tracers. There's a big large Q array, of dimensions (ncols,ntracers,nlevs), containing all the tracers. But there are different kind of tracers, and some may require certain updates that others don't. E.g., shoc needs to do apply some sort of turbulent scheme to some of them, but not all; radiation, on the other hand, needs some of them for optical property stuff. All the atm procs that use some (but not all) of the tracers can benefit from having their subset of tracers contiguous within Q. Here, contiguous refers to the tracer index, so that their subset of tracers looks like `Q(:, my_start:my_end, :)`, which is not contiguous in memory, but once subviewed at a particular col idx is.

So, in order to accommodate (if possible) all requests for contiguous sub-groups, I need to rearrange the order of the tracers in Q, in a way that all the sub-groups are a contiguous set of slices. This utility is doing this: given some sets of objects (of the same type, variable sizes), find a way to list the elements in their union that allows each of these sets to be a contiguous subset of it.

Reply:
Note: I'm using `list` instead of `set` in the impl, for code reasons, not 'conceptual' reasons. From the conceptual point of view, those lists are like sets (meaning, no repeated entries, and the order of elements doesn't matter). But I can control how entries in a list are ordered, while set does not let me rearrange them.

Reply:
Cool. I was confused before because I thought that the existence of this function was an assertion that such an ordering could always be found, but I see now that it returns an empty list if this is not the case. And using a `list` does indeed make sense if you want control of the ordering, since `set` is an ordered set and `unordered_set` is indifferent to ordering.

If elements in a group have a canonical ordering, I think `contiguous_superset` is a decent name. To me, It doesn't really seem like a partitioning, since partitions are disjoint and the function returns a single list from which overlapping subsets may be selected.

My only question is: are there circumstances of interest to SCREAM where one could find no such ordering that would cause trouble for the caller? Another way to ask this question is: does the possibility of no such ordering place an undue burden on the caller to handle edge cases, or is this unusual in practice?

Reply:
Honestly, I don't know. I think there might be such cases once we get aerosols. Also, SL transport in Homme, as I understand, will distinguish between active and passive tracers.

If we ever end up with a set of tracers that does not allow all the subgroups to be contiguous, and if these subgroups _really need_ to be bundled, then we will need some re-thinking of the impl of the atm procs issuing those requests.

I believe if that happens, we will have to pick 1+ of these procs, and turn its request for bundled fields into a "preferred" bundling. The FM does a few tries when building bundled groups: first, all the 'Required' and 'Preferred' bundling requests are considered; if that fails, it starts removing one 'Preferred' request at a time, until it succeeds (which it will, at the latest when there's only one request left). So if we ever hit the case where all atm procs issue a GroupRequest with 'Required' bundling, we'll have to go through 1+ of these atm procs, and change the request to 'Preferred'. However, that means that the atm proc will have to be able to handle a contiguous bundle (like `Q(:,start:end,:)`) as well as completely independent fields.

So, long story short, I think it's possible, as the number of parametrizations grows, that we'll hit a case where we can't bundle all groups. If that happens, 1+ of the parametrizations will have to implement a "fall-back" strategy, where it can deal with separate fields rather than a single N+1 dimensional field.

Reply:
Tbc, I don't think we should implement the 'fall-back' strategy now. Homme and SHOC (the only current users of groups) can still issue a GroupRequest with bundling=Required. We'll cross that bridge if/when we get there.

Reply:
Another note: Homme currently is adding a constraint, namely that qv must be the 1st tracer. There really isn't a need for that, since accessing a single tracer at iq=0 is the same as accessing at iq=23. At some point, I'd like to change homme, to store `qv_idx`, and using that every time it needs qv (which is needed for T<->T_virt conversions). However, that's a non-trivial change in Homme, since it might be time-consuming to track down all the places where Q is accessed only for qv's values. So for now I opted for hacking the FM, so that if qv is present _and_ the "tracers" group is present, we manage to put qv at the front of the list.

Forcing qv at the front _might_ be a constraint that causes the FM to fail to honor all requests. I don't think it will, but it's worth keeping this "pointless" constraint in mind.

Note: the logic for putting qv at the front of the list might fail if there is a group G that is a superset of the tracers. I don't think that will happen any time soon, but it's worth keeping in mind (I think I added a comment in the code).